File: android/Reguerta/app/src/main/java/com/reguerta/user/data/freshness/FirestoreCriticalDataFreshnessRemoteRepository.kt
package com.reguerta.user.data.freshness

import com.google.firebase.Timestamp
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Source
import com.reguerta.user.data.firestore.ReguertaFirestoreCollection
import com.reguerta.user.data.firestore.ReguertaFirestoreDocument
import com.reguerta.user.data.firestore.ReguertaFirestoreEnvironment
import com.reguerta.user.data.firestore.ReguertaFirestorePath
import com.reguerta.user.data.firestore.toRepositoryException
import com.reguerta.user.domain.RepositoryErrorKind
import com.reguerta.user.domain.RepositoryException
import com.reguerta.user.domain.freshness.CriticalCollection
import com.reguerta.user.domain.freshness.CriticalDataFreshnessConfig
import com.reguerta.user.domain.freshness.CriticalDataFreshnessRemoteRepository
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.tasks.await

class FirestoreCriticalDataFreshnessRemoteRepository(
    private val firestore: FirebaseFirestore,
) : CriticalDataFreshnessRemoteRepository {
    override suspend fun getConfig(environment: String): CriticalDataFreshnessConfig {
        val resolvedEnvironment = ReguertaFirestoreEnvironment.entries.firstOrNull { candidate ->
            candidate.wireValue == environment
        } ?: throw invalidCriticalDataFreshnessConfig()
        val memberConfigDocumentPath = ReguertaFirestorePath(environment = resolvedEnvironment).documentPath(
            collection = ReguertaFirestoreCollection.CONFIG,
            documentId = ReguertaFirestoreDocument.MEMBER.wireValue,
        )
        try {
            val snapshot = firestore.document(memberConfigDocumentPath).get(Source.SERVER).await()
            if (!snapshot.exists()) {
                throw RepositoryException(
                    kind = RepositoryErrorKind.NOT_FOUND,
                    resource = CRITICAL_DATA_FRESHNESS_RESOURCE,
                )
            }

            val cacheExpiration = snapshot.getLong("cacheExpirationMinutes")
                ?.takeIf { it in 1..Int.MAX_VALUE.toLong() }
                ?: throw invalidCriticalDataFreshnessConfig()
            val timestampMap = snapshot.get("lastTimestamps") as? Map<*, *>
                ?: throw invalidCriticalDataFreshnessConfig()
            val remoteTimestamps = CriticalCollection.entries.associateWith { collection ->
                val timestampMillis = when (val rawValue = timestampMap[collection.wireKey]) {
                    is Timestamp -> rawValue.toDate().time
                    is java.util.Date -> rawValue.time
                    else -> throw invalidCriticalDataFreshnessConfig()
                }
                timestampMillis.takeIf { it > 0L }
                    ?: throw invalidCriticalDataFreshnessConfig()
            }

            return CriticalDataFreshnessConfig(
                cacheExpirationMinutes = cacheExpiration.toInt(),
                remoteTimestampsMillis = remoteTimestamps,
            )
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            throw error.toRepositoryException(resource = CRITICAL_DATA_FRESHNESS_RESOURCE)
        }
    }
}

private const val CRITICAL_DATA_FRESHNESS_RESOURCE = "config.member.freshness"

private fun invalidCriticalDataFreshnessConfig() = RepositoryException(
    kind = RepositoryErrorKind.INVALID_DATA,
    resource = CRITICAL_DATA_FRESHNESS_RESOURCE,
)

File: ios/Reguerta/Reguerta/Data/Freshness/FirestoreCriticalDataFreshnessRemoteRepository.swift
import FirebaseCore
import FirebaseFirestore
import Foundation

actor FirestoreCriticalDataFreshnessRemoteRepository: CriticalDataFreshnessRemoteRepository {
    private let storedDB: Firestore

    init(firebaseAppName: String) {
        guard let app = FirebaseApp.app(name: firebaseAppName) else {
            preconditionFailure("Firebase app is required for critical-data freshness")
        }
        self.storedDB = Firestore.firestore(app: app)
    }

    func getConfig(environment: SessionEnvironment) async throws -> CriticalDataFreshnessConfig {
        do {
            let snapshot = try await storedDB
                .reguertaDocument(.memberConfiguration, in: .config, environment: environment)
                .getDocument(source: .server)

            guard snapshot.exists else { throw RepositoryError.notFound(resource: "config.member") }
            guard let data = snapshot.data() else { throw RepositoryError.invalidData(resource: "config.member") }
            return try Self.config(data: data)
        } catch {
            throw FirestoreRepositoryErrorMapper.map(error, resource: "config.member")
        }
    }

    nonisolated static func config(data: [String: Any]) throws -> CriticalDataFreshnessConfig {
        guard let cacheExpirationNumber = data["cacheExpirationMinutes"] as? NSNumber,
              CFGetTypeID(cacheExpirationNumber) != CFBooleanGetTypeID(),
              cacheExpirationNumber.doubleValue.rounded() == cacheExpirationNumber.doubleValue,
              cacheExpirationNumber.int64Value > 0,
              cacheExpirationNumber.int64Value <= Int64(Int.max),
              let lastTimestamps = data["lastTimestamps"] as? [String: Any]
        else {
            throw RepositoryError.invalidData(resource: "config.member")
        }

        var remoteTimestamps: [CriticalCollection: Int64] = [:]
        for collection in CriticalCollection.allCases {
            guard let timestamp = lastTimestamps[collection.rawValue] as? Timestamp else {
                throw RepositoryError.invalidData(resource: "config.member.lastTimestamps")
            }
            let milliseconds = Int64(timestamp.dateValue().timeIntervalSince1970 * 1_000)
            guard milliseconds > 0 else { throw RepositoryError.invalidData(resource: "config.member.lastTimestamps") }
            remoteTimestamps[collection] = milliseconds
        }

        return CriticalDataFreshnessConfig(
            cacheExpirationMinutes: cacheExpirationNumber.intValue,
            remoteTimestampsMillis: remoteTimestamps
        )
    }
}
